### PR TITLE
Add simplified APIs for model obtaining maxtext models

### PR DESCRIPTION
**Pull request by A9isha**
# Description

Adding a simplified API of `from_pretrained()` which can work minimally as

For creating all the necessary configs and mesh for RL, we can use the APIs.

```
import maxtext as mt

config= mt.pyconfig(model_name="llama3.1-8b-Instruct")
model,mesh = mt.from_pretrained(config)

#OR
config= mt.pyconfig(model_name="llama3.1-8b-Instruct")
## your own mesh
model = mt.from_pretrained(config, mesh)


# OR for train_rl.py
trainer_config, sampler_config, trainer_devices, sampler_devices = mt.setup_configs_and_devices(model_name="llama3.1-8b-Instruct")

reference_model, reference_mesh, actor_model, actor_mesh, rollout_mesh = mt.create_models_and_meshes(
      trainer_config, sampler_config, trainer_devices, sampler_devices
  )

# OR regular invocation of train_rl.py

run_name=maz-8b-$RANDOM python3 -m src.maxtext.trainers.post_train.rl.train_rl \
 model_name=llama3.1-8b-Instruct run_name=$run_name\
    steps=4 rollout_tensor_parallelism=-1\
 rollout_data_parallelism=1 rollout_expert_parallelism=1\
 test_batch_start_index=10  num_test_batches=15 
base_output_directory=/path/to/say/gcs/bucket

# OR standalone script where you want to invoke 

    MAXTEXT_CONFIGS_DIR = "src/maxtext/configs"
    import maxtext as mt

    config= mt.pyconfig(model_name=MAXTEXT_MODEL_VERSION, 
    hf_access_token=args.hf_token,
    base_output_directory="gs://path/to/save/artifacts",
    base_config=f"{MAXTEXT_CONFIGS_DIR}/post_train/rl.yml",
    )
    qwen2_actor, _= mt.from_pretrained(config) # ref_mesh and train_mesh are the same for us
  else:
    qwen2_actor = params_lib.create_model_from_safe_tensors(
        MODEL_PATH, config, trainer_mesh, dtype=MODEL_DTYPE
    )

```

FIXES: b/492376313

# Tests

Ran locally

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/pyconfig.py [hunk 1]
```diff
@@ -77,16 +77,45 @@ def _module_from_path(path: str) -> str | None:
...
+def _resolve_or_infer_config(argv: list[str] | None = None, **kwargs) -> tuple[str, list[str]]:
   """Resolves or infers config file path from module."""
+  if argv is None:
+    argv = [""]
+
+  if kwargs.get("base_config"):
+    logger.info("Using config : %s", kwargs["base_config"])
+    return resolve_config_path(kwargs["base_config"]), argv[1:] if len(argv) > 1 else []
+
   if len(argv) >= 2 and argv[1].endswith(".yml"):
```
Comment (wang2yn84) [inline src/maxtext/configs/pyconfig.py:92]: There are certain restriction of the first 2 args passed to the API, can you document this?

File: src/maxtext/utils/model_creation_utils.py [hunk 2]
```diff
@@ -258,8 +261,171 @@ def _create_model(rng_key=None):
...
+  reference_model, reference_mesh = from_pretrained(trainer_config, devices=trainer_devices, wrap_with_tunix_adapter=True)
+  devices_array = maxtext_utils.create_device_mesh(sampler_config, sampler_devices)
+  rollout_mesh = Mesh(devices_array, sampler_config.mesh_axes)
+
+  if trainer_config.load_checkpoint_only_once:
+    max_logging.log("Creating policy model by copying reference model instead of restoring from checkpoint again.")
+    with reference_mesh:
+      actor_base_model = nnx.clone(reference_model.base)
+      use_no_op_mappings = "maxtext_config" in trainer_config.vllm_additional_config
+      actor_model = TunixMaxTextAdapter(base_model=actor_base_model, use_no_op_mappings=use_no_op_mappings)
```
Comment (wang2yn84) [inline src/maxtext/utils/model_creation_utils.py:379]: Can you also document with Tunix adapter, how to add weight mapping if there are new models?

**Review by wang2yn84 — APPROVED** [review]
Thank you very much!

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py [hunk 3]
```diff
@@ -251,7 +251,7 @@ def load_weights(self, rng_key: jax.Array) -> None:
       return
 
     with self.mesh, nn.logical_axis_rules(""):
-      model, _ = model_creation_utils.create_nnx_model(
+      model, _ = model_creation_utils.from_pretrained(
```
Comment (NicoGrande) [inline src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: Can the second output be removed here with the new API?

File: src/maxtext/utils/model_creation_utils.py [hunk 4]
```diff
@@ -258,8 +261,171 @@ def _create_model(rng_key=None):
   return _create_model_partial, abstract_model
 
 
-def create_nnx_model(config, mesh=None, devices=None, model_mode=MODEL_MODE_TRAIN, rng_key=None):
+def setup_configs_and_devices(argv: list[str] | None = None, kwargs: dict | None = None, **extra_kwargs):
```
Comment (NicoGrande) [inline src/maxtext/utils/model_creation_utils.py:266]: I understand that we are trying to simplify train_rl, but IIUC this function is only useful for RL? Does it make sense to move it to this more general utility? Perhaps we can rename it to make this clear.

File: src/maxtext/utils/model_creation_utils.py [hunk 5]
```diff
@@ -258,8 +261,171 @@ def _create_model(rng_key=None):
...
+    trainer_config = pyconfig.initialize_pydantic(argv, **trainer_kwargs)
+    sampler_config = pyconfig.initialize_pydantic(argv, **sampler_kwargs)
+
+  else:
+    raise ValueError("num_trainer_slices and num_samplers_slices should be both -1 or positive")
+
+  return trainer_config, sampler_config, trainer_devices, sampler_devices
+
+
+def create_models_and_meshes(trainer_config, sampler_config, trainer_devices, sampler_devices):
```
Comment (NicoGrande) [inline src/maxtext/utils/model_creation_utils.py:360]: Same as above.

Comment (NicoGrande) [reply src/maxtext/utils/model_creation_utils.py:360]: Also if we are refactoring here, I think its a good opportunity to add some typing and docstrings to these methods :)

File: src/maxtext/utils/model_creation_utils.py [hunk 6]
```diff
@@ -258,8 +261,171 @@ def _create_model(rng_key=None):
...
+      # Try to convert checkpoint on the fly
+      if not config.hf_access_token:
+        raise ValueError("hf_access_token must be provided when not providing a pre-existing checkpoint")
+
+      max_logging.warning("Checkpoint path is not provided, converting checkpoint to orbax format for MaxText")
+
+      simulated_cpu_devices_count = 16
+
+      import subprocess  # pylint: disable=import-outside-toplevel
+      import sys  # pylint: disable=import-outside-toplevel
```
Comment (NicoGrande) [inline src/maxtext/utils/model_creation_utils.py]: Why not move imports to top level?

File: src/maxtext/utils/model_creation_utils.py [hunk 7]
```diff
@@ -258,8 +261,171 @@ def _create_model(rng_key=None):
...
+  original_mesh = mesh
+  if config.convert_checkpoint_if_possible:
+    if not (epath.Path(config.base_output_directory) / "0" / "items").exists():
+      # Try to convert checkpoint on the fly
+      if not config.hf_access_token:
+        raise ValueError("hf_access_token must be provided when not providing a pre-existing checkpoint")
+
+      max_logging.warning("Checkpoint path is not provided, converting checkpoint to orbax format for MaxText")
+
+      simulated_cpu_devices_count = 16
```
Comment (NicoGrande) [inline src/maxtext/utils/model_creation_utils.py]: Does this depend on the environment at all? What does hardcoding to 16 here mean?

File: src/maxtext/configs/pyconfig.py (hunk 1 above)
Comment (A9isha) [reply src/maxtext/configs/pyconfig.py:92]: done

File: src/maxtext/utils/model_creation_utils.py (hunk 6 above)
Comment (A9isha) [reply src/maxtext/utils/model_creation_utils.py]: done

File: src/maxtext/utils/model_creation_utils.py (hunk 5 above)
Comment (A9isha) [reply src/maxtext/utils/model_creation_utils.py:360]: done

File: src/maxtext/utils/model_creation_utils.py (hunk 7 above)
Comment (A9isha) [reply src/maxtext/utils/model_creation_utils.py]: discussed offline, adding a comment to explain

File: src/maxtext/utils/model_creation_utils.py (hunk 4 above)
Comment (A9isha) [reply src/maxtext/utils/model_creation_utils.py:266]: discussed offline, added a doc string explaining RL applicability

File: src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py (hunk 3 above)
Comment (A9isha) [reply src/maxtext/integration/vllm/maxtext_vllm_adapter/adapter.py]: done - thanks!

File: src/maxtext/utils/model_creation_utils.py (hunk 2 above)
Comment (A9isha) [reply src/maxtext/utils/model_creation_utils.py:379]: done - thanks Lance

File: src/maxtext/configs/pyconfig.py [hunk 8]
```diff
@@ -77,16 +77,48 @@ def _module_from_path(path: str) -> str | None:
...
 
-def _resolve_or_infer_config(argv: list[str]) -> tuple[str, list[str]]:
+def _resolve_or_infer_config(argv: list[str] | None = None, **kwargs) -> tuple[str, list[str]]:
   """Resolves or infers config file path from module."""
+  if argv is None:
+    argv = [""]
+
+  if kwargs.get("base_config"):
+    logger.info("Using config : %s", kwargs["base_config"])
+    return resolve_config_path(kwargs["base_config"]), argv[1:] if len(argv) > 1 else []
```
Comment (igorts-git) [inline src/maxtext/configs/pyconfig.py]: Nit. `argv[1:]` is sufficient, no need for the if-else block.
```py
$ python3
>>> a = []
>>> a[1:]
[]
>>> b = ["a"]
>>> b[1:]
[]
```

File: src/maxtext/configs/pyconfig.py [hunk 9]
```diff
@@ -77,16 +77,48 @@ def _module_from_path(path: str) -> str | None:
...
-    raise ValueError(f"No config file provided and no default config found for module '{module}'")
-  config_path = os.path.join(MAXTEXT_CONFIGS_DIR, _CONFIG_FILE_MAPPING[module])
-  logger.warning("No config file provided, using default config mapping: %s", config_path)
-  return config_path, argv[1:]
+    config_path = os.path.join(MAXTEXT_CONFIGS_DIR, "base.yml")
+    logger.warning("No config file provided and no default config found for module '%s', using base.yml", module)
+  else:
+    config_path = os.path.join(MAXTEXT_CONFIGS_DIR, _CONFIG_FILE_MAPPING[module])
+    logger.warning("No config file provided, using default config mapping: %s", config_path)
+  remaining_argv = argv[1:] if len(argv) > 1 else []
```
Comment (igorts-git) [inline src/maxtext/configs/pyconfig.py]: same here. Just use `argv[1:]`

File: src/maxtext/utils/model_creation_utils.py [hunk 10]
```diff
@@ -258,8 +263,181 @@ def _create_model(rng_key=None):
...
+
+def from_pretrained(
+    config, mesh=None, devices=None, model_mode=MODEL_MODE_TRAIN, rng_key=None, wrap_with_tunix_adapter=False
+):
   """Creates a NNX model with sharded parameters, possibly loading from a checkpoint."""
+  original_mesh = mesh
+  if config.convert_checkpoint_if_possible:
+    if not (epath.Path(config.base_output_directory) / "0" / "items").exists():
+      # Try to convert checkpoint on the fly
+      if not config.hf_access_token:
```
Comment (igorts-git) [inline src/maxtext/utils/model_creation_utils.py:397]: should we first check `load_parameters_path` before attempting to convert the checkpoint? Maybe the user already converted it and specified in the command line.

Comment (A9isha) [reply src/maxtext/utils/model_creation_utils.py:397]: excellent point - thank you Igor!

File: src/maxtext/configs/pyconfig.py (hunk 9 above)
Comment (A9isha) [reply src/maxtext/configs/pyconfig.py]: done - thanks

File: src/maxtext/configs/pyconfig.py (hunk 8 above)
Comment (A9isha) [reply src/maxtext/configs/pyconfig.py]: done - thank you Igor
